### PR TITLE
Add IndexShard#getLatestReplicationCheckpoint behind segrep enable feature flag

### PR DESCRIPTION
**Pull request by dreamer-89**
Signed-off-by: Suraj Singh <surajrider@gmail.com>

### Description
Came from discussion in https://github.com/opensearch-project/OpenSearch/pull/4041#discussion_r939084536, this PR adds `getLatestReplicationCheckpoint` behind feature flag. 
 
### Issues Related
#2212 #3988 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/test/java/org/opensearch/index/shard/IndexShardTests.java [hunk 1]
```diff
@@ -3500,6 +3501,27 @@ public void testReadSnapshotConcurrently() throws IOException, InterruptedExcept
         closeShards(newShard);
     }
 
+    /**
+     *  Test that latestReplicationCheckpoint returns null only for docrep enabled indices
+     */
+    public void testReplicationCheckpointNullForDocRep() throws IOException {
```
Comment (mch2) [inline server/src/test/java/org/opensearch/index/shard/IndexShardTests.java:3507]: nit - pls move to SegmentReplicationIndexShardTests.

**Review by mch2 — APPROVED** [review]
LGTM with nit.

File: server/src/test/java/org/opensearch/index/shard/IndexShardTests.java (hunk 1 above)
Comment (dreamer-89) [reply server/src/test/java/org/opensearch/index/shard/IndexShardTests.java:3507]: Done

**Comment by dreamer-89** [comment]
The gradle check looks like have a legit failure

```
REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testReplicationOnDone" -Dtests.seed=F2650EF01D035C3B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=it -Dtests.timezone=Asia/Aqtobe -Druntime.java=17

org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testReplicationOnDone FAILED
    java.lang.NullPointerException: Cannot invoke "org.opensearch.indices.replication.checkpoint.ReplicationCheckpoint.getShardId()" because "this.initialCheckpoint" is null
        at __randomizedtesting.SeedInfo.seed([F2650EF01D035C3B:E8ABDE3DD7FFB399]:0)
        at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.setUp(SegmentReplicationTargetServiceTests.java:66)

REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testShardAlreadyReplicating" -Dtests.seed=F2650EF01D035C3B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=it -Dtests.timezone=Asia/Aqtobe -Druntime.java=17

org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testShardAlreadyReplicating FAILED
    java.lang.NullPointerException: Cannot invoke "org.opensearch.indices.replication.checkpoint.ReplicationCheckpoint.getShardId()" because "this.initialCheckpoint" is null
        at __randomizedtesting.SeedInfo.seed([F2650EF01D035C3B:D1BC8BEB97C49532]:0)
        at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.setUp(SegmentReplicationTargetServiceTests.java:66)

REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testNewCheckpoint_validationPassesAndReplicationFails" -Dtests.seed=F2650EF01D035C3B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=it -Dtests.timezone=Asia/Aqtobe -Druntime.java=17

org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testNewCheckpoint_validationPassesAndReplicationFails FAILED
    java.lang.NullPointerException: Cannot invoke "org.opensearch.indices.replication.checkpoint.ReplicationCheckpoint.getShardId()" because "this.initialCheckpoint" is null
        at __randomizedtesting.SeedInfo.seed([F2650EF01D035C3B:D84FBDC9FED9E1BA]:0)
        at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.setUp(SegmentReplicationTargetServiceTests.java:66)

REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testAlreadyOnNewCheckpoint" -Dtests.seed=F2650EF01D035C3B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=it -Dtests.timezone=Asia/Aqtobe -Druntime.java=17

org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testAlreadyOnNewCheckpoint FAILED
    java.lang.NullPointerException: Cannot invoke "org.opensearch.indices.replication.checkpoint.ReplicationCheckpoint.getShardId()" because "this.initialCheckpoint" is null
        at __randomizedtesting.SeedInfo.seed([F2650EF01D035C3B:B0BE1A7DA4F7BCC7]:0)
        at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.setUp(SegmentReplicationTargetServiceTests.java:66)

REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testNewCheckpointBehindCurrentCheckpoint" -Dtests.seed=F2650EF01D035C3B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=it -Dtests.timezone=Asia/Aqtobe -Druntime.java=17

org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testNewCheckpointBehindCurrentCheckpoint FAILED
    java.lang.NullPointerException: Cannot invoke "org.opensearch.indices.replication.checkpoint.ReplicationCheckpoint.getShardId()" because "this.initialCheckpoint" is null
        at __randomizedtesting.SeedInfo.seed([F2650EF01D035C3B:68A6B7BA36E97637]:0)
        at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.setUp(SegmentReplicationTargetServiceTests.java:66)

REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testBeforeIndexShardClosed_CancelsOngoingReplications" -Dtests.seed=F2650EF01D035C3B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=it -Dtests.timezone=Asia/Aqtobe -Druntime.java=17

org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testBeforeIndexShardClosed_CancelsOngoingReplications FAILED
    java.lang.NullPointerException: Cannot invoke "org.opensearch.indices.replication.checkpoint.ReplicationCheckpoint.getShardId()" because "this.initialCheckpoint" is null
        at __randomizedtesting.SeedInfo.seed([F2650EF01D035C3B:800E2FC6951F71F0]:0)
        at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.setUp(SegmentReplicationTargetServiceTests.java:66)
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.gradle.api.internal.tasks.testing.worker.TestWorker (file:/home/ubuntu/.gradle/wrapper/dists/gradle-7.5-all/6qsw290k5lz422uaf8jf6m7co/gradle-7.5/lib/plugins/gradle-testing-****-7.5.jar)
WARNING: Please consider reporting this to the maintainers of org.gradle.api.internal.tasks.testing.worker.TestWorker
WARNING: System::setSecurityManager will be removed in a future release

REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testTargetThrowsException" -Dtests.seed=F2650EF01D035C3B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=it -Dtests.timezone=Asia/Aqtobe -Druntime.java=17

org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testTargetThrowsException FAILED
    java.lang.NullPointerException: Cannot invoke "org.opensearch.indices.replication.checkpoint.ReplicationCheckpoint.getShardId()" because "this.initialCheckpoint" is null
        at __randomizedtesting.SeedInfo.seed([F2650EF01D035C3B:3BFFE9A7800C979B]:0)
        at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.setUp(SegmentReplicationTargetServiceTests.java:66)
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.gradle.api.internal.tasks.testing.worker.TestWorker (file:/home/ubuntu/.gradle/wrapper/dists/gradle-7.5-all/6qsw290k5lz422uaf8jf6m7co/gradle-7.5/lib/plugins/gradle-testing-****-7.5.jar)
WARNING: Please consider reporting this to the maintainers of org.gradle.api.internal.tasks.testing.worker.TestWorker
WARNING: System::setSecurityManager will be removed in a future release

REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testTargetReturnsSuccess_listenerCompletes" -Dtests.seed=F2650EF01D035C3B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=it -Dtests.timezone=Asia/Aqtobe -Druntime.java=17

org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testTargetReturnsSuccess_listenerCompletes FAILED
    java.lang.NullPointerException: Cannot invoke "org.opensearch.indices.replication.checkpoint.ReplicationCheckpoint.getShardId()" because "this.initialCheckpoint" is null
        at __randomizedtesting.SeedInfo.seed([F2650EF01D035C3B:F9DE9955F8C69EA6]:0)
        at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.setUp(SegmentReplicationTargetServiceTests.java:66)

REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationSourceHandlerTests.testSendFiles_emptyRequest" -Dtests.seed=F2650EF01D035C3B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=ja-JP -Dtests.timezone=Etc/GMT+2 -Druntime.java=17

org.opensearch.indices.replication.SegmentReplicationSourceHandlerTests > testSendFiles_emptyRequest FAILED
    java.lang.AssertionError
        at __randomizedtesting.SeedInfo.seed([F2650EF01D035C3B:F50229272AF394A]:0)
        at org.junit.Assert.fail(Assert.java:87)
        at org.junit.Assert.fail(Assert.java:96)
        at org.opensearch.indices.replication.SegmentReplicationSourceHandlerTests$2.onFailure(SegmentReplicationSourceHandlerTests.java:124)
        at org.opensearch.action.ActionRunnable.onFailure(ActionRunnable.java:103)
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:54)
        at org.opensearch.common.util.concurrent.OpenSearchExecutors$DirectExecutorService.execute(OpenSearchExecutors.java:341)
        at org.opensearch.common.util.concurrent.ListenableFuture.notifyListener(ListenableFuture.java:120)
        at org.opensearch.common.util.concurrent.ListenableFuture.lambda$done$0(ListenableFuture.java:112)
        at java.****/java.util.ArrayList.forEach(ArrayList.java:1511)
        at org.opensearch.common.util.concurrent.ListenableFuture.done(ListenableFuture.java:112)
        at org.opensearch.common.util.concurrent.BaseFuture.setException(BaseFuture.java:178)
        at org.opensearch.common.util.concurrent.ListenableFuture.onFailure(ListenableFuture.java:149)
        at org.opensearch.indices.replication.SegmentReplicationSourceHandler.lambda$sendFiles$1(SegmentReplicationSourceHandler.java:111)
        at org.opensearch.core.internal.io.IOUtils.close(IOUtils.java:87)
        at org.opensearch.core.internal.io.IOUtils.closeWhileHandlingException(IOUtils.java:177)
        at org.opensearch.core.internal.io.IOUtils.closeWhileHandlingException(IOUtils.java:167)
        at org.opensearch.core.internal.io.IOUtils.closeWhileHandlingException(IOUtils.java:155)
        at org.opensearch.indices.replication.SegmentReplicationSourceHandler.lambda$sendFiles$2(SegmentReplicationSourceHandler.java:111)
        at org.opensearch.action.ActionListener$1.onFailure(ActionListener.java:88)
        at org.opensearch.action.ActionListener$1.onResponse(ActionListener.java:82)
        at org.opensearch.common.util.concurrent.ListenableFuture$1.doRun(ListenableFuture.java:126)
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
        at org.opensearch.common.util.concurrent.OpenSearchExecutors$DirectExecutorService.execute(OpenSearchExecutors.java:341)
        at org.opensearch.common.util.concurrent.ListenableFuture.notifyListener(ListenableFuture.java:120)
        at org.opensearch.common.util.concurrent.ListenableFuture.addListener(ListenableFuture.java:82)
        at org.opensearch.action.StepListener.whenComplete(StepListener.java:93)
        at org.opensearch.indices.replication.SegmentReplicationSourceHandler.sendFiles(SegmentReplicationSourceHandler.java:144)
        at org.opensearch.indices.replication.SegmentReplicationSourceHandlerTests.testSendFiles_emptyRequest(SegmentReplicationSourceHandlerTests.java:115)
```

File: server/src/main/java/org/opensearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -1396,10 +1396,13 @@ public GatedCloseable<IndexCommit> acquireSafeIndexCommit() throws EngineExcepti
     }
 
     /**
-     * Returns the lastest Replication Checkpoint that shard received. Shards will return an EMPTY checkpoint before
-     * the engine is opened.
+     * Returns the latest ReplicationCheckpoint that shard received.
+     * @return EMPTY checkpoint before the engine is opened and null for non-segrep enabled indices
      */
     public ReplicationCheckpoint getLatestReplicationCheckpoint() {
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/shard/IndexShard.java:1402]: Should we change the signature to Optional<ReplicationCheckpoint> as well?

Comment (dreamer-89) [reply server/src/main/java/org/opensearch/index/shard/IndexShard.java:1402]: >Should we change the signature to Optional as well?

Thanks @Bukhtawar for the review. Did you mean updating java doc to mention this method is applicable only for SegRep enabled indices ?

**Comment by dreamer-89** [comment]
Last gradle job fail with below, which is not failing on local. Refiring the gradle check
```
REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testReplicationOnDone" -Dtests.seed=7736EA6F094D816D -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=bg-BG -Dtests.timezone=Africa/Dakar -Druntime.java=17

org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testReplicationOnDone FAILED
    org.mockito.exceptions.verification.TooFewActualInvocations: 
    segmentReplicationTargetService.onNewCheckpoint(
        ReplicationCheckpoint{shardId=[index][0], primaryTerm=85, segmentsGen=2, seqNo=-1, version=7},
        <any>
    );
    Wanted 2 times:
    -> at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testReplicationOnDone(SegmentReplicationTargetServiceTests.java:240)
    But was 1 time:
    -> at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testReplicationOnDone(SegmentReplicationTargetServiceTests.java:234)
        at __randomizedtesting.SeedInfo.seed([7736EA6F094D816D:6DF83AA2C3B16ECF]:0)
        at app//org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testReplicationOnDone(SegmentReplicationTargetServiceTests.java:240)`REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testReplicationOnDone" -Dtests.seed=7736EA6F094D816D -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=bg-BG -Dtests.timezone=Africa/Dakar -Druntime.java=17

org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testReplicationOnDone FAILED
    org.mockito.exceptions.verification.TooFewActualInvocations: 
    segmentReplicationTargetService.onNewCheckpoint(
        ReplicationCheckpoint{shardId=[index][0], primaryTerm=85, segmentsGen=2, seqNo=-1, version=7},
        <any>
    );
    Wanted 2 times:
    -> at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testReplicationOnDone(SegmentReplicationTargetServiceTests.java:240)
    But was 1 time:
    -> at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testReplicationOnDone(SegmentReplicationTargetServiceTests.java:234)
        at __randomizedtesting.SeedInfo.seed([7736EA6F094D816D:6DF83AA2C3B16ECF]:0)
        at app//org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testReplicationOnDone(SegmentReplicationTargetServiceTests.java:240)
```

**Pull request merged**